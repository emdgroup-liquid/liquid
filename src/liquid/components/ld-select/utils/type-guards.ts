export const isLdOption = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionElement | HTMLLdOptionInternalElement =>
  ["LD-OPTION", "LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroup = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupElement | HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP", "LD-OPTGROUP-INTERNAL"].includes(
    (el as HTMLElement)?.tagName,
  );

export const isLdOptionInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionInternalElement =>
  ["LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroupInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP-INTERNAL"].includes((el as HTMLElement)?.tagName);

type HTMLLdOptInternal =
  | HTMLLdOptionInternalElement
  | HTMLLdOptgroupInternalElement;
export const isLdOptInternalHidden = (
  opt: HTMLLdOptInternal,
): opt is
  | (HTMLLdOptInternal & {
      hidden: true;
    })
  | (HTMLLdOptInternal & {
      filtered: true;
    }) => {
  return opt.hidden || opt.filtered;
};
