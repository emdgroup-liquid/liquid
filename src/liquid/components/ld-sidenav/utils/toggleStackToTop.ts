import { LdSidenavNavitem } from '../ld-sidenav-navitem/ld-sidenav-navitem'

export const toggleStackToTop = (el: HTMLElement, stacked: boolean) => {
  const outerSlider = el.closest('ld-sidenav-slider')
  if (!outerSlider) return

  const sidenavPaddingY =
    parseInt(
      window
        .getComputedStyle(outerSlider)
        .getPropertyValue('--ld-sidenav-padding-y')
    ) * 16

  // If not stacked, put everything back in place.
  if (!stacked) {
    el.style.removeProperty('--ld-slider-navitem-move-up')
    return
  }

  // Else, calculate how much vertical space is taken by stacked items above
  // and move the nav item up accordingly.
  let totalSpaceAbove = 0
  let totalSpaceOccupiedAbove = 0
  for (const elem of Array.from(el.parentElement.children)) {
    if (el === elem) break

    if (
      elem.tagName === 'LD-SIDENAV-SEPARATOR' ||
      (elem.tagName === 'LD-SIDENAV-NAVITEM' &&
        !['secondary', 'tertiary'].includes(
          (elem as unknown as LdSidenavNavitem).mode
        ))
    ) {
      totalSpaceOccupiedAbove += elem.getBoundingClientRect().height
    } else {
      totalSpaceAbove += sidenavPaddingY
    }
    totalSpaceAbove += elem.getBoundingClientRect().height
  }

  const spaceToMoveUp = totalSpaceAbove - totalSpaceOccupiedAbove
  el.style.setProperty('--ld-slider-navitem-move-up', `-${spaceToMoveUp}px`)
}
