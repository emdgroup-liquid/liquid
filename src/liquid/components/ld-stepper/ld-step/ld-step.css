.ld-step {
  --ld-step-description-font-size: var(--ld-step-description-font-size-md);
  --ld-step-description-font-size-sm: 0.625rem;
  --ld-step-description-font-size-md: 0.75rem;
  --ld-step-description-font-size-lg: 0.875rem;
  --ld-step-dot-offset: calc(
    (var(--ld-step-dot-size-with-icon) - var(--ld-step-dot-size)) / 2
  );
  --ld-step-dot-size: var(--ld-step-dot-size-default);
  --ld-step-dot-size-default: var(--ld-step-dot-size-md);
  --ld-step-dot-size-sm: var(--ld-sp-12);
  --ld-step-dot-size-md: var(--ld-sp-16);
  --ld-step-dot-size-lg: 1.25rem;
  --ld-step-dot-size-with-icon: var(--ld-step-dot-size-with-icon-md);
  --ld-step-dot-size-with-icon-sm: 1.25rem;
  --ld-step-dot-size-with-icon-md: var(--ld-sp-24);
  --ld-step-dot-size-with-icon-lg: var(--ld-sp-32);
  --ld-step-icon-size: calc(var(--ld-step-dot-size-with-icon) * 0.5);
  --ld-step-icon-size-custom: calc(var(--ld-step-dot-size-with-icon) * 0.6);
  --ld-step-icon-offset: calc(
    (var(--ld-step-dot-size-with-icon) - var(--ld-step-icon-size)) / 2
  );
  --ld-step-label-font-size: var(--ld-step-label-font-size-md);
  --ld-step-label-font-size-sm: 0.75rem;
  --ld-step-label-font-size-md: 0.875rem;
  --ld-step-label-font-size-lg: 1.125rem;
  --ld-step-line-thickness: var(--ld-step-line-thickness-md);
  --ld-step-line-thickness-sm: var(--ld-sp-2);
  --ld-step-line-thickness-md: 0.1875rem;
  --ld-step-line-thickness-lg: var(--ld-sp-4);
  --ld-step-text-distance: var(--ld-sp-16);
  --ld-step-text-margin: calc(
    var(--ld-step-dot-size-with-icon) + var(--ld-step-text-distance)
  );
  --ld-step-vertical-label-margin-top: calc(
    /* remove actual line-height of the label from dot size to get the difference */
      (var(--ld-step-dot-size-with-icon) - var(--ld-step-label-font-size) * 1.4) /
      2
  );
  --ld-step-vertical-text-gap: var(--ld-step-vertical-text-gap-md);
  --ld-step-vertical-text-gap-sm: var(--ld-sp-4);
  --ld-step-vertical-text-gap-md: var(--ld-sp-6);
  --ld-step-vertical-text-gap-lg: var(--ld-sp-8);

  /* colors */
  --ld-step-bg-col: var(--ld-thm-primary);
  --ld-step-col: var(--ld-step-col-default);
  --ld-step-col-default: var(--ld-col-neutral-400);
  --ld-step-col-done: var(--ld-thm-primary);
  --ld-step-dot-col: var(--ld-step-dot-col-default);
  --ld-step-dot-col-default: var(--ld-col-neutral-200);
  --ld-step-dot-col-current: var(--ld-thm-secondary);
  --ld-step-dot-col-done: var(--ld-thm-primary);
  --ld-step-description-col: var(--ld-step-description-col-default);
  --ld-step-description-col-default: var(--ld-step-col);
  --ld-step-description-col-done: var(--ld-col-neutral-900);
  --ld-step-icon-col: var(--ld-col-wht);
  --ld-step-line-start-col: var(--ld-step-line-col-default);
  --ld-step-line-end-col: var(--ld-step-line-col-default);
  --ld-step-line-col-default: var(--ld-col-neutral-200);
  --ld-step-line-col-done: var(--ld-thm-primary);

  /* on-brand colors */
  --ld-step-brand-col: var(--ld-col-wht-alpha-highest);
  --ld-step-brand-col-done: var(--ld-col-wht);
  --ld-step-description-brand-col: var(--ld-col-wht-alpha-highest);
  --ld-step-description-brand-col-done: var(--ld-col-wht);
  --ld-step-dot-brand-col: var(--ld-thm-primary-focus);
  --ld-step-dot-brand-col-done: var(--ld-col-wht);
  --ld-step-brand-icon-col: var(--ld-thm-primary);
  --ld-step-line-brand-col: var(--ld-thm-primary-focus);
  --ld-step-line-brand-col-done: var(--ld-col-wht);

  align-items: center;
  display: flex;
  flex-direction: column;
  height: 100%;
  position: relative;

  a,
  button {
    align-items: center;
    background: transparent;
    border: 0;
    color: var(--ld-step-col);
    display: flex;
    flex-direction: column;
    font-family: var(--ld-font-body);
    font-size: var(--ld-step-label-font-size);
    font-weight: 700;
    line-height: 140%;
    margin: var(--ld-step-text-margin) 0 0; /* Safari reset */
    padding: 0;
    text-align: center;

    &::before {
      background: var(--ld-step-dot-col);
      border-radius: var(--ld-br-full);
      box-sizing: border-box;
      content: '';
      display: block;
      height: var(--ld-step-dot-size);
      position: absolute;
      top: var(--ld-step-dot-offset);
      width: var(--ld-step-dot-size);
      z-index: 1;
    }
  }

  .ld-icon {
    --ld-icon-size-sm: var(--ld-step-icon-size);
    --ld-icon-size-md: var(--ld-step-icon-size);
    --ld-icon-size-lg: var(--ld-step-icon-size);
    color: var(--ld-step-icon-col);
    pointer-events: none;
    position: absolute;
    top: var(--ld-step-icon-offset);
    z-index: 1;
  }

  &--brand-color {
    --ld-step-col-default: var(--ld-step-brand-col);
    --ld-step-col-done: var(--ld-step-brand-col-done);
    --ld-step-description-col-default: var(--ld-step-description-brand-col);
    --ld-step-description-col-done: var(--ld-step-description-brand-col-done);
    --ld-step-dot-col-default: var(--ld-step-dot-brand-col);
    --ld-step-dot-col-done: var(--ld-step-dot-brand-col-done);
    --ld-step-icon-col: var(--ld-step-brand-icon-col);
    --ld-step-line-col-default: var(--ld-step-line-brand-col);
    --ld-step-line-col-done: var(--ld-step-line-brand-col-done);
  }

  &--custom-icon {
    --ld-step-icon-size: var(--ld-step-icon-size-custom);
  }

  &--with-icon {
    --ld-step-dot-size: var(--ld-step-dot-size-with-icon);
  }

  &--current {
    --ld-step-col: var(--ld-step-col-done);
    --ld-step-description-col: var(--ld-step-description-col-done);
    --ld-step-dot-col: var(--ld-step-dot-col-current);
    --ld-step-line-start-col: var(--ld-step-line-col-done);
  }

  &--done {
    --ld-step-col: var(--ld-step-col-done);
    --ld-step-description-col: var(--ld-step-description-col-done);
    --ld-step-dot-col: var(--ld-step-dot-col-done);
    --ld-step-line-start-col: var(--ld-step-line-col-done);
    --ld-step-line-end-col: var(--ld-step-line-col-done);

    a,
    button {
      cursor: pointer;
    }
  }

  &--optional {
    --ld-step-icon-col: var(--ld-step-dot-col);

    a,
    button {
      &::before {
        background: transparent;
        border: solid var(--ld-step-line-thickness) var(--ld-step-dot-col);
      }
    }
  }

  &--skipped {
    --ld-step-col: var(--ld-step-col-done);
    --ld-step-description-col: var(--ld-step-description-col-done);
    --ld-step-dot-col: var(--ld-step-dot-col-done);
    --ld-step-line-start-col: var(--ld-step-line-col-done);
    --ld-step-line-end-col: var(--ld-step-line-col-done);

    a,
    button {
      cursor: pointer;
    }
  }

  &--vertical {
    align-items: start;
    gap: var(--ld-step-vertical-text-gap);
    padding: 0;
    text-align: left;

    a,
    button {
      margin: var(--ld-step-vertical-label-margin-top) 0 0
        var(--ld-step-text-margin);

      &::before {
        flex-shrink: 0;
        left: calc(var(--ld-step-dot-size-with-icon) / 2);
        transform: translateX(-50%);
      }
    }

    .ld-icon {
      left: var(--ld-step-icon-offset);
    }
  }

  :host(:empty) & a,
  :host(:empty) & button,
  a:empty,
  button:empty {
    margin: var(--ld-step-dot-offset) 0 0;

    &,
    &::before {
      position: relative;
      top: 0;
    }
  }

  &--sm {
    --ld-step-description-font-size: var(--ld-step-description-font-size-sm);
    --ld-step-dot-size-default: var(--ld-step-dot-size-sm);
    --ld-step-dot-size-with-icon: var(--ld-step-dot-size-with-icon-sm);
    --ld-step-label-font-size: var(--ld-step-label-font-size-sm);
    --ld-step-line-thickness: var(--ld-step-line-thickness-sm);
    --ld-step-vertical-text-gap: var(--ld-step-vertical-text-gap-sm);
  }

  &--lg {
    --ld-step-description-font-size: var(--ld-step-description-font-size-lg);
    --ld-step-dot-size-default: var(--ld-step-dot-size-lg);
    --ld-step-dot-size-with-icon: var(--ld-step-dot-size-with-icon-lg);
    --ld-step-label-font-size: var(--ld-step-label-font-size-lg);
    --ld-step-line-thickness: var(--ld-step-line-thickness-lg);
    --ld-step-vertical-text-gap: var(--ld-step-vertical-text-gap-lg);
  }
}

:host(:not(:first-of-type)) .ld-step,
.ld-step:not(:first-of-type) {
  &::before {
    content: '';
    background: var(--ld-step-line-start-col);
    height: var(--ld-step-line-thickness);
    left: calc(var(--ld-stepper-gap) / -2);
    position: absolute;
    right: calc(50% + var(--ld-step-dot-size) / 2 - var(--ld-sp-1));
    top: calc(var(--ld-step-dot-size-with-icon) / 2);
    transform: translateY(-50%);
  }

  &.ld-step--vertical::before {
    bottom: calc(100% - var(--ld-step-dot-offset) - var(--ld-sp-1));
    height: auto;
    left: calc(var(--ld-step-dot-size-with-icon) / 2);
    right: auto;
    top: calc(var(--ld-stepper-gap) / -2);
    transform: translatex(-50%);
    width: var(--ld-step-line-thickness);
  }
}

:host(:not(:last-of-type)) .ld-step,
.ld-step:not(:last-of-type) {
  &::after {
    content: '';
    background: var(--ld-step-line-end-col);
    height: var(--ld-step-line-thickness);
    left: calc(50% + var(--ld-step-dot-size) / 2 - var(--ld-sp-1));
    position: absolute;
    right: calc(var(--ld-stepper-gap) / -2);
    top: calc(var(--ld-step-dot-size-with-icon) / 2);
    transform: translateY(-50%);
    z-index: -1;
  }

  &.ld-step--vertical::after {
    bottom: calc(var(--ld-stepper-gap) / -2);
    height: auto;
    left: calc(var(--ld-step-dot-size-with-icon) / 2);
    right: auto;
    top: calc(
      var(--ld-step-dot-size) + var(--ld-step-dot-offset) - var(--ld-sp-1)
    );
    transform: translatex(-50%);
    width: var(--ld-step-line-thickness);
  }
}

.ld-step__description {
  color: var(--ld-step-description-col);
  font-family: var(--ld-font-body);
  font-size: var(--ld-step-description-font-size);
  line-height: 140%;
  margin-left: var(--ld-step-text-margin);
}
