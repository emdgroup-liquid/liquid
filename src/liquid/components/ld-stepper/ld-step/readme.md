---
eleventyNavigation:
  key: Step
  parent: Stepper
layout: layout.njk
title: Step
permalink: components/ld-stepper/ld-step/
---

# ld-step

The `ld-step` component is a subcomponent for `ld-stepper`.

Please refer to the [`ld-stepper` documentation](components/ld-stepper) for usage examples.

---


<!-- Auto Generated Below -->


## Properties

| Property           | Attribute            | Description                                                                         | Type               | Default          |
| ------------------ | -------------------- | ----------------------------------------------------------------------------------- | ------------------ | ---------------- |
| `brandColor`       | `brand-color`        | Switch colors for brand background                                                  | `boolean`          | `false`          |
| `current`          | `current`            | Step is the current step                                                            | `boolean`          | `false`          |
| `description`      | `description`        | Description text to display below the step name (vertical mode only)                | `string`           | `undefined`      |
| `done`             | `done`               | Step is done                                                                        | `boolean`          | `false`          |
| `icon`             | `icon`               | Permanently show a custom icon inside the dot                                       | `string`           | `undefined`      |
| `key`              | `key`                | for tracking the node's identity when working with lists                            | `string \| number` | `undefined`      |
| `labelCurrent`     | `label-current`      | Label for current step (scree-reader only)                                          | `string`           | `'Current'`      |
| `labelDone`        | `label-done`         | Label for step that is done (scree-reader only)                                     | `string`           | `'Done'`         |
| `labelOptional`    | `label-optional`     | Label for step that is optional (scree-reader only)                                 | `string`           | `'Optional'`     |
| `labelSkipped`     | `label-skipped`      | Label for step that was skipped (scree-reader only)                                 | `string`           | `'Skipped'`      |
| `labelWasOptional` | `label-was-optional` | Additional hint in label for step that is done and was optional (scree-reader only) | `string`           | `'was optional'` |
| `ldTabindex`       | `ld-tabindex`        | Tab index of the step                                                               | `number`           | `undefined`      |
| `optional`         | `optional`           | Step may be skipped                                                                 | `boolean`          | `false`          |
| `ref`              | `ref`                | reference to component                                                              | `any`              | `undefined`      |
| `size`             | `size`               | Step size                                                                           | `"lg" \| "sm"`     | `undefined`      |
| `skipped`          | `skipped`            | Step was skipped                                                                    | `boolean`          | `false`          |
| `vertical`         | `vertical`           | Vertical layout                                                                     | `boolean`          | `false`          |


## Events

| Event            | Description | Type                                             |
| ---------------- | ----------- | ------------------------------------------------ |
| `ldstepselected` |             | `CustomEvent<{ index: number; label: string; }>` |


## Methods

### `focusInner() => Promise<void>`

Sets focus on the step

#### Returns

Type: `Promise<void>`




## Shadow Parts

| Part            | Description                                  |
| --------------- | -------------------------------------------- |
| `"button"`      | actual `button` element                      |
| `"description"` | `span` element wrapping the description text |
| `"focusable"`   |                                              |
| `"li"`          | actual `li` element                          |


## Dependencies

### Depends on

- [ld-sr-only](../../ld-sr-only)
- [ld-icon](../../ld-icon)

### Graph
```mermaid
graph TD;
  ld-step --> ld-sr-only
  ld-step --> ld-icon
  style ld-step fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

 
