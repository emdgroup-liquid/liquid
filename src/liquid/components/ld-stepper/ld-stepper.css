:host,
.ld-stepper {
  --ld-stepper-gap: var(--ld-sp-16);

  display: block;

  ol {
    display: grid;
    grid-auto-columns: minmax(0, 1fr);
    grid-auto-flow: column;
    gap: var(--ld-stepper-gap);
    padding: 0;
  }
}

.ld-stepper--vertical {
  :host(&),
  & {
    ol {
      flex-direction: column;
      grid-auto-flow: row;
      grid-auto-rows: 1fr;
    }
  }
}

.ld-stepper--fit-content {
  :host(&),
  & {
    ol {
      display: flex;
    }
  }
}
