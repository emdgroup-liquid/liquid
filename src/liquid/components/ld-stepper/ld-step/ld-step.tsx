import {
  Component,
  Host,
  h,
  Prop,
  Element,
  Event,
  EventEmitter,
  Method,
} from '@stencil/core'
import { getClassNames } from 'src/liquid/utils/getClassNames'

export type SelectedDetail = { index: number; label: string }

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part description - `span` element wrapping the description text
 * @part button - actual `button` element
 * @part li - actual `li` element
 */
@Component({
  tag: 'ld-step',
  styleUrl: 'ld-step.css',
  shadow: true,
})
export class LdStep implements InnerFocusable {
  @Element() el: HTMLLdStepElement
  private button: HTMLButtonElement

  /** Switch colors for brand background */
  @Prop() brandColor = false
  /** Step is the current step */
  @Prop() current = false
  /** Description text to display below the step name (vertical mode only) */
  @Prop() description: string
  /** Step is done */
  @Prop() done = false
  /** Permanently show a custom icon inside the dot */
  @Prop() icon?: HTMLLdIconElement['name']
  /** Tab index of the step */
  @Prop() ldTabindex: number | undefined
  /** Step may be skipped */
  @Prop() optional = false
  /** Step was skipped */
  @Prop() skipped = false
  /** Vertical layout */
  @Prop() vertical = false

  @Event() ldstepselected: EventEmitter<SelectedDetail>

  /** Sets focus on the step */
  @Method()
  async focusInner() {
    this.button?.focus()
  }

  handleClick = () => {
    this.ldstepselected.emit({
      index: Array.from(this.el.parentElement.children).findIndex(
        (child) => child === this.el
      ),
      label: this.el.textContent,
    })
  }

  componentDidLoad() {
    if (this.current) {
      this.handleClick()
    }
  }

  render() {
    return (
      <Host>
        <li
          class={getClassNames([
            'ld-step',
            this.brandColor && 'ld-step--brand-color',
            this.current && 'ld-step--current',
            this.done && 'ld-step--done',
            this.icon && 'ld-step--custom-icon',
            (this.done || this.icon) && 'ld-step--with-icon',
            this.optional && 'ld-step--optional',
            this.skipped && 'ld-step--skipped',
            this.vertical && 'ld-step--vertical',
          ])}
          part="li"
          role="listitem"
        >
          <button
            aria-current={this.current ? 'step' : undefined}
            disabled={!this.done && !this.skipped}
            onClick={this.handleClick}
            part="button focusable"
            ref={(ref) => (this.button = ref)}
            tabIndex={this.ldTabindex}
            type="button"
          >
            <slot></slot>
          </button>
          {(this.done || this.icon) && (
            <ld-icon name={this.icon ?? 'checkmark'} />
          )}
          {this.description && this.vertical && (
            <span class="ld-step__description" part="description">
              {this.description}
            </span>
          )}
        </li>
      </Host>
    )
  }
}
