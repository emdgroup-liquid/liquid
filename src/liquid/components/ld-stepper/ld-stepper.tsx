import { Component, Element, Host, h, Prop, State, Watch } from '@stencil/core'
import { getClassNames } from 'src/liquid/utils/getClassNames'
import { SelectedDetail } from './ld-step/ld-step'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part list - `ol` element wrapping the slot
 * @part label - `ld-sr-only` element containing the label
 */
@Component({
  tag: 'ld-stepper',
  styleUrl: 'ld-stepper.css',
  shadow: true,
})
export class LdStepper {
  @Element() el: HTMLLdStepperElement

  /** Switch colors for brand background. */
  @Prop() brandColor = false
  /** Indicates whether the steps should be evenly distributed or fit to their content */
  @Prop() fitContent = false
  /** Template for the screen-reader label, containing the label of the current step and the steps summary */
  @Prop() labelTemplate = '$1, $2'
  /** Step summary template for the screen-reader label, containing the index of the current step and the overall number of steps */
  @Prop() labelSummaryTemplate = 'step $1 of $2'
  /** Step size */
  @Prop() size?: HTMLLdStepElement['size']
  /** Vertical layout */
  @Prop() vertical = false

  @State() currentLabel: string
  @State() currentIndex: number
  @State() steps: NodeListOf<HTMLLdStepElement>

  private getLabel() {
    const summary = this.labelSummaryTemplate
      .replace('$1', String(this.currentIndex + 1))
      .replace('$2', String(this.steps.length))

    return this.currentLabel
      ? this.labelTemplate
          .replace('$1', this.currentLabel)
          .replace('$2', summary)
      : summary
  }

  updateCurrent = (event: CustomEvent<SelectedDetail>) => {
    this.currentIndex = event.detail.index
    this.currentLabel = event.detail.label
  }

  @Watch('brandColor')
  @Watch('size')
  @Watch('vertical')
  private propagateProps() {
    this.steps.forEach((ldStep) => {
      ldStep.brandColor = this.brandColor
      ldStep.size = this.size
      ldStep.vertical = this.vertical
    })
  }

  componentWillLoad() {
    this.steps = this.el.querySelectorAll('ld-step')

    this.propagateProps()
  }

  render() {
    return (
      <Host
        class={getClassNames([
          'ld-stepper',
          this.fitContent && 'ld-stepper--fit-content',
          this.size && `ld-stepper--${this.size}`,
          this.vertical && 'ld-stepper--vertical',
        ])}
        role="navigation"
        onLdstepselected={this.updateCurrent}
      >
        <ld-sr-only>{this.getLabel()}</ld-sr-only>
        <ol part="list">
          <slot></slot>
        </ol>
      </Host>
    )
  }
}
