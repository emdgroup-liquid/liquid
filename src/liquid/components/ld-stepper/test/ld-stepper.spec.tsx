import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdStep } from '../ld-step/ld-step'
import { LdStepper } from '../ld-stepper'
import '../../../utils/mutationObserver'

describe('ld-stepper', () => {
  describe('renders', () => {
    it('default', async () => {
      const page = await newSpecPage({
        components: [LdStepper, LdStep],
        template: () => (
          <ld-stepper>
            <ld-step current />
          </ld-stepper>
        ),
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with size prop', async () => {
      const page = await newSpecPage({
        components: [LdStepper, LdStep],
        template: () => (
          <ld-stepper size="sm">
            <ld-step current />
          </ld-stepper>
        ),
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with current step label', async () => {
      const page = await newSpecPage({
        components: [LdStepper, LdStep],
        template: () => (
          <ld-stepper>
            <ld-step current>Step 1</ld-step>
          </ld-stepper>
        ),
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with fit-content', async () => {
      const page = await newSpecPage({
        components: [LdStepper, LdStep],
        template: () => (
          <ld-stepper fitContent>
            <ld-step current />
          </ld-stepper>
        ),
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with custom label', async () => {
      const page = await newSpecPage({
        components: [LdStepper, LdStep],
        template: () => (
          <ld-stepper
            labelTemplate="$2: $1"
            labelSummaryTemplate="Schritt $1 von $2"
          >
            <ld-step current>Step 1</ld-step>
          </ld-stepper>
        ),
      })
      expect(page.root).toMatchSnapshot()
    })

    it('vertical', async () => {
      const page = await newSpecPage({
        components: [LdStepper, LdStep],
        template: () => (
          <ld-stepper vertical>
            <ld-step current />
          </ld-stepper>
        ),
      })
      expect(page.root).toMatchSnapshot()
    })

    it('on brand color', async () => {
      const page = await newSpecPage({
        components: [LdStepper, LdStep],
        template: () => (
          <ld-stepper brandColor>
            <ld-step current />
          </ld-stepper>
        ),
      })
      expect(page.root).toMatchSnapshot()
    })
  })

  it('propagates prop changes', async () => {
    const page = await newSpecPage({
      components: [LdStepper, LdStep],
      template: () => (
        <ld-stepper>
          <ld-step current />
        </ld-stepper>
      ),
    })
    const ldStepper = page.root as HTMLLdStepperElement

    ldStepper.brandColor = true
    ldStepper.size = 'lg'
    ldStepper.vertical = true
    await page.waitForChanges()

    expect(page.root).toMatchSnapshot()
  })
})
