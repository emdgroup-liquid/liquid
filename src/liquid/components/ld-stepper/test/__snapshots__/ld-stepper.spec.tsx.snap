// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-stepper propagates prop changes 1`] = `
<ld-stepper class="ld-stepper ld-stepper--lg ld-stepper--vertical" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      step 1 of 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--brand-color ld-step--current ld-step--lg ld-step--vertical" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
  </ld-step>
</ld-stepper>
`;

exports[`ld-stepper renders default 1`] = `
<ld-stepper class="ld-stepper" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      step 1 of 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--current" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
  </ld-step>
</ld-stepper>
`;

exports[`ld-stepper renders on brand color 1`] = `
<ld-stepper class="ld-stepper" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      step 1 of 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--brand-color ld-step--current" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
  </ld-step>
</ld-stepper>
`;

exports[`ld-stepper renders vertical 1`] = `
<ld-stepper class="ld-stepper ld-stepper--vertical" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      step 1 of 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--current ld-step--vertical" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
  </ld-step>
</ld-stepper>
`;

exports[`ld-stepper renders with current step label 1`] = `
<ld-stepper class="ld-stepper" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      Step 1, step 1 of 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--current" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
    Step 1
  </ld-step>
</ld-stepper>
`;

exports[`ld-stepper renders with custom label 1`] = `
<ld-stepper class="ld-stepper" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      Schritt 1 von 1: Step 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--current" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
    Step 1
  </ld-step>
</ld-stepper>
`;

exports[`ld-stepper renders with fit-content 1`] = `
<ld-stepper class="ld-stepper ld-stepper--fit-content" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      step 1 of 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--current" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
  </ld-step>
</ld-stepper>
`;

exports[`ld-stepper renders with size prop 1`] = `
<ld-stepper class="ld-stepper ld-stepper--sm" role="navigation">
  <mock:shadow-root>
    <ld-sr-only>
      step 1 of 1
    </ld-sr-only>
    <ol part="list">
      <slot></slot>
    </ol>
  </mock:shadow-root>
  <ld-step>
    <mock:shadow-root>
      <li class="ld-step ld-step--current ld-step--sm" part="li" role="listitem">
        <ld-sr-only>
          Current:
        </ld-sr-only>
        <button aria-current="step" part="button focusable" type="button">
          <slot></slot>
        </button>
      </li>
    </mock:shadow-root>
  </ld-step>
</ld-stepper>
`;
